xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc

```diff
@@ -0,0 +1,1708 @@
+  // this is for assertions/diagnostics only.
+  [[maybe_unused]] bool is_proc_;
+
+  // Maps id -> MLIR value for all SSA values inside the current
+  // function base (function or proc body).
+  std::unordered_map<int64_t, Value> ssa_map_;
+
+  // Maps name -> MLIR value for all state elements of this proc.
+  std::unordered_map<std::string, Value> state_elem_map;
+};
+
+//===----------------------------------------------------------------------===//
+// Type Translation
+//===----------------------------------------------------------------------===//
+
+Type translateType(::xls::Type* xls_type, OpBuilder& builder,
+                   MLIRContext* ctx) {
+  switch (xls_type->kind()) {
+    case ::xls::TypeKind::kTuple: {
+      std::vector<Type> types{};
+      for (auto type : xls_type->AsTupleOrDie()->element_types()) {
+        types.push_back(translateType(type, builder, ctx));
+      }
+      return TupleType::get(ctx, types);
+    }
+    case ::xls::TypeKind::kBits: {
+      return builder.getIntegerType(xls_type->AsBitsOrDie()->bit_count());
+    }
+    case ::xls::TypeKind::kArray: {
+      return ArrayType::get(
+          xls_type->AsArrayOrDie()->size(),
+          translateType(xls_type->AsArrayOrDie()->element_type(), builder,
+                        ctx));
+    }
+    case ::xls::TypeKind::kToken: {
+      return TokenType::get(ctx);
+    }
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Literal Translation
+//===----------------------------------------------------------------------===//
+
+absl::StatusOr<Operation*> translateBitsLiteral(const ::xls::Bits& b,
+                                                Location loc,
+                                                OpBuilder& builder,
+                                                MLIRContext* ctx,
+                                                TranslationState& state) {
+  APInt converted_value;
+
+  if (b.bit_count() == 0) {
+    converted_value = APInt(/*numBits=*/0, /*val=*/0, /*isSigned=*/false,
+                            /*implicitTrunc=*/false);
+  } else {
+    uint64_t num_words = b.bitmap().word_count();
+    std::vector<uint64_t> words{};
+    words.reserve(num_words);
+    for (uint64_t i = 0; i < num_words; i++) {
+      words.push_back(b.bitmap().GetWord(i));
+    }
+    converted_value = APInt(b.bit_count(), words);
+  }
+
+  auto type = builder.getIntegerType(b.bit_count());
+  return builder
+      .create<ConstantScalarOp>(loc, type,
+                                builder.getIntegerAttr(type, converted_value))
+      .getOperation();
+}
+
+absl::StatusOr<Operation*> translateLiteral(const ::xls::Value& b,
+                                            OpBuilder& builder,
+                                            MLIRContext* ctx,
+                                            TranslationState& state) {
+  switch (b.kind()) {
+    case ::xls::ValueKind::kBits:
```

Comment:
I'd suggest adding `using ::xls::ValueKind;` at the top of this namespace to make these switches easier to read.